Issue 85

vadekh opened the pull request:
**Related Issue:**
<!-- This project only accepts pull requests related to open issues. -->
<!-- Please open an issue for this change if one does not already exist. -->
closes #85 

## Description
<!-- In 1-3 sentences, provide an overview of what changes were made and why. -->
Created a ISO 639-3 field in the language editor. The field has no functionality at the moment.
Created a testing file for LanguageEditor.js
## Checklist

- [x] Check out the [Contributing Guidelines][contributing] if you need help getting started with your pull request.

- [x] [Open an issue][issues] for the change (if one doesn't already exist).

- [x] Update tests for planned changes. Check that they are failing. (`npm test`)

- [x] Write code to pass the tests, then build the project. (`npm run build`)

- [x] Check that tests are now passing (`npm test`). This will also build the project again.

- [x] Add inline code commenting (using [JSDoc][JSDoc] style).

- [x] Generate the developer documentation (`npm run docs`) and check your changes by opening `docs/index.html` in a browser.

- [x] If you're adding a new component, add yourself as a [codeowner][codeowners] for that component.

[codeowners]:   https://github.com/digitallinguistics/app/blob/master/.github/CODEOWNERS
[contributing]: https://github.com/digitallinguistics/app/blob/master/.github/CONTRIBUTING.md
[issues]:       https://github.com/digitallinguistics/app/issues
[JSDoc]:        https://jsdoc.app/
[Node]:         https://nodejs.org/en/
[nvm]:          https://github.com/creationix/nvm
[nvm-windows]:  https://github.com/coreybutler/nvm-windows

vadekh commented:
@dwhieb I just copied the html that you suggested. I would love some feedback on the test file though. Not sure if that was thorough enough, or too thorough, etc. I also proactively added a `beforeEach` section to the test which will be required once the scope of the tests goes beyond rendering.

dwhieb commented on src/components/LanguageEditor/LanguageEditor.test.js:
```diff
@@ -0,0 +1,24 @@
+describe(`LanguageEditor`, () => {
+
+  beforeEach( () => {
+    cy.visit(`/#/langauges`);
```
I do this _all_ the time. I regularly search projects for this typo out of habit now 😄.

```suggestion
    cy.visit(`/#/languages`);
```

dwhieb commented on src/components/LanguageEditor/LanguageEditor.test.js:
```diff
@@ -0,0 +1,24 @@
+  beforeEach( () => {
+    cy.visit(`/#/langauges`);
+    cy.get(`.language-dropdown`)
+    .select(`add`);
```
If you wanted to bypass the dependency on the Language Dropdown, another way you might structure the test is this:

```js
// retrieve test data from the /test/fixtures/languages.json file
cy.fixture(`/languages.json`)
// add test data to the database (local storage for the time being)
.then(languages => localStorage.setItem(`languages`, JSON.stringify(languages)));

// navigate to the Languages page for the test language with the ID "1" (Spanish)
cy.visit(`/#/languages/1`);

// the assertions for the test
// ...
```

Structuring the test this way means that there's one less piece of the app that your test depends on, keeping your test focused on what you intend to be testing. The way the test is currently written, it would break if the Language Dropdown stopped working. That's fine, because it alerts us to the problem, but the tests for the Language Dropdown would already do that. So ideally this test will _only_ break if there's a problem with the Language Editor (usually not actually possible to set up tests this way in E2E testing, but it's a principle to go by).

dwhieb commented on src/components/LanguageEditor/LanguageEditor.test.js:
```diff
@@ -0,0 +1,24 @@
+    .contains(`p`, `The ISO 69-3 code for this language variety.`)
+    .should(`be.visible`)
+    .contains(`Search for a language code here.`)
+    .should(`have.attr`, `href`, `https://iso639-3.sil.org/code_tables/639/data`);
```
I think we can remove these assertions, because they're about static content rather than functionality. They're a little like testing for "the banner should be blue", which isn't so much functionality as just a design choice. The text around the ISO field is similarly just a content / design choice.

We _do_ want to test that the ISO field renders at all though. So you should add a test to look for `.language-editor input[name=iso]`. We probably don't want to use a selector like `.language-editor > label > div > input[name=iso]`, because those other pieces of HTML aren't really central to the functionality we're interested in here (the ISO field itself).

dwhieb commented on src/components/LanguageEditor/LanguageEditor.test.js:
```diff
@@ -0,0 +1,24 @@
+    .should(`have.class`, `language-editor`)
+
+    //Check for ISO 639-3 Code
+    .contains(`ISO 639-3 Code`)
```
I like that you chained `.contains()` off the Language Editor. I would have just done a new `cy.get('input[name=iso]')`, but yours actually checks to confirm that the field is contained within the Language Editor.

If the code starts to become hard to read (so far it's fine), you could do `cy.get('.language-editor input[name=iso]')`, which would also check that the ISO field is contained within the Language Editor.

vadekh commented:
@dwhieb Thanks for the feedback. I simplified the testing to focus on functional elements. I also added the `cy.within()` function to test for the ISO field (and others in the future) inside the LanguageEditor form.

dwhieb commented:
@vadekh Nice! Didn't think about using `cy.within()` - good thinking. This is good to go!
